ros_ws/src/autonomous/neuralnetwork/src/ai_trainer.cpp

```diff
@@ -0,0 +1,246 @@
+    }
+    m_gen = 0;
+    ROS_INFO_STREAM("created first generation randomly with " + std::to_string(m_generation_size) + " entities");
+}
+
+// TODO
+void AiTrainer::chooseBestFromGeneration()
+{
+    for (int i = 0; i < m_parameter_training_generation_best; i++)
+    {
+        int best_net = -1;
+        for (int j = 0; j < m_generation_size; j++)
+        {
+            if (best_net == -1 || m_scores[j] > m_scores[best_net])
+            {
+                best_net = j;
+            }
+        }
+        m_best_nets[i] = m_nets[best_net];
+        m_scores[best_net] = -1;
+    }
+    // save best
+    m_best_nets[0]->save(m_parameter_config_folder + "/best_of/champion_" + std::to_string(m_gen) + ".config");
+}
+
+void AiTrainer::createNextGeneration()
+{
+    chooseBestFromGeneration();
+    for (int i = 0, big_i = 0; i < m_parameter_training_generation_best; i++)
+    {
+        // parent
+        FANN::neural_net* parent = m_best_nets[i];
+
+        // copy parent
+        m_nets[big_i] = parent;
+        big_i++;
+
+        // create parent mutations
+        for (int i = 0; i < m_parameter_training_generation_multiplier; i++)
+        {
+            FANN::neural_net* child = new FANN::neural_net();
+            child->create_standard_array(NUM_LAYERS, NET_ARGS);
+            cloneNet(child, parent);
+            mutate(child, m_parameter_training_learning_rate);
+            m_nets[big_i] = child;
+            big_i++;
+        }
+    }
+
+    // reset scores
+    for (int i = 0; i < m_generation_size; i++)
+    {
+        m_scores[i] = 0;
+    }
+    m_gen++;
+}
+
+void AiTrainer::cloneNet(FANN::neural_net* to, FANN::neural_net* from)
+{
+    long size = to->get_total_connections();
+
+    FANN::connection c_from[size];
+    from->get_connection_array(c_from);
+
+    to->set_weight_array(c_from, size);
+}
+
+void AiTrainer::mutate(FANN::neural_net* net, fann_type rate)
+{
+    long size = net->get_total_connections();
+    FANN::connection connections[size];
+    net->get_connection_array(connections);
+
+    // creating random vector
+    std::vector<fann_type> mutation_delta = random_vector(size, rate, r_mult_one);
+
+    // applying vector
+    for (int i = 0; i < size; i++)
+    {
+        connections[i].weight = connections[i].weight * mutation_delta[i];
+    }
+}
+
+void AiTrainer::lapTimerCallback(const ros::TimerEvent&)
+{
+    update();
+}
+
+void AiTrainer::deploy(FANN::neural_net* net)
+{
+    net->randomize_weights(-1, 1);
+
+    neuralnetwork::net_param message;
+    // size
+    long size = net->get_total_connections();
+    message.size = size;
+    // connections
+    FANN::connection connections[size];
+    net->get_connection_array(connections);
+    std::vector<unsigned int> from_neurons;
+    std::vector<unsigned int> to_neurons;
+    std::vector<float> weights;
+    for (int i = 0; i < size; i++)
+    {
+        weights.push_back(connections[i].weight);
+    }
+    message.weights = weights;
```

Comment:
You could fill the `message.weights` vector directly which would prevent unnecessary copies e.g.
```C++
    message.weights.clear();
    for (size_t i = 0; i < size; i++)
    {
        message.weights.push_back(connections[i].weight);
    }
```